File: web/src/lib/console-summary.ts
import { resolveTunnelStatus } from '@/lib/tunnel-model';
import type { Client, ConsoleSummary } from '@/types';

export const EMPTY_CONSOLE_SUMMARY: ConsoleSummary = {
  total_clients: 0,
  online_clients: 0,
  offline_clients: 0,
  total_tunnels: 0,
  active_tunnels: 0,
  inactive_tunnels: 0,
  pending_tunnels: 0,
  offline_tunnels: 0,
  stopped_tunnels: 0,
  error_tunnels: 0,
};

export function summarizeConsoleClients(clients: Client[] | null | undefined): ConsoleSummary {
  const base: ConsoleSummary = { ...EMPTY_CONSOLE_SUMMARY };

  for (const client of clients ?? []) {
    base.total_clients += 1;
    if (client.online) {
      base.online_clients += 1;
    } else {
      base.offline_clients += 1;
    }

    for (const tunnel of client.proxies ?? []) {
      base.total_tunnels += 1;

      switch (resolveTunnelStatus(tunnel, client.online).key) {
        case 'exposed':
          base.active_tunnels += 1;
          break;
        case 'pending':
          base.pending_tunnels += 1;
          break;
        case 'offline':
          base.offline_tunnels += 1;
          break;
        case 'stopped':
          base.stopped_tunnels += 1;
          break;
        case 'error':
          base.error_tunnels += 1;
          break;
      }
    }
  }

  base.inactive_tunnels = base.total_tunnels - base.active_tunnels;
  return base;
}

File: internal/server/console_summary.go
package server

import "netsgo/pkg/protocol"

type consoleSummaryView struct {
	TotalClients    int `json:"total_clients"`
	OnlineClients   int `json:"online_clients"`
	OfflineClients  int `json:"offline_clients"`
	TotalTunnels    int `json:"total_tunnels"`
	ActiveTunnels   int `json:"active_tunnels"`
	InactiveTunnels int `json:"inactive_tunnels"`
	PendingTunnels  int `json:"pending_tunnels"`
	OfflineTunnels  int `json:"offline_tunnels"`
	PausedTunnels   int `json:"paused_tunnels"`
	StoppedTunnels  int `json:"stopped_tunnels"`
	ErrorTunnels    int `json:"error_tunnels"`
}

type consoleData struct {
	Clients []clientView
	Summary consoleSummaryView
}

func (s *Server) collectConsoleData() consoleData {
	clients := s.collectClientViews()
	return consoleData{
		Clients: clients,
		Summary: summarizeConsoleClients(clients),
	}
}

func summarizeConsoleClients(clients []clientView) consoleSummaryView {
	summary := consoleSummaryView{}
	for _, client := range clients {
		summary.TotalClients++
		if client.Online {
			summary.OnlineClients++
		} else {
			summary.OfflineClients++
		}

		for _, tunnel := range client.Proxies {
			summary.TotalTunnels++
			switch consoleTunnelStatusKey(tunnel, client.Online) {
			case protocol.ProxyRuntimeStateExposed:
				summary.ActiveTunnels++
			case protocol.ProxyRuntimeStatePending:
				summary.PendingTunnels++
			case protocol.ProxyRuntimeStateOffline:
				summary.OfflineTunnels++
			case "paused":
				summary.PausedTunnels++
			case "stopped":
				summary.StoppedTunnels++
			default:
				summary.ErrorTunnels++
			}
		}
	}
	summary.InactiveTunnels = summary.TotalTunnels - summary.ActiveTunnels
	return summary
}

func consoleTunnelStatusKey(tunnel protocol.ProxyConfig, clientOnline bool) string {
	runtimeState := tunnel.RuntimeState
	if !clientOnline && tunnel.DesiredState == protocol.ProxyDesiredStateRunning && runtimeState != protocol.ProxyRuntimeStateError {
		runtimeState = protocol.ProxyRuntimeStateOffline
	}

	switch runtimeState {
	case protocol.ProxyRuntimeStatePending,
		protocol.ProxyRuntimeStateExposed,
		protocol.ProxyRuntimeStateOffline,
		protocol.ProxyRuntimeStateError:
		return runtimeState
	case protocol.ProxyRuntimeStateIdle:
		if tunnel.DesiredState == protocol.ProxyDesiredStatePaused {
			return "paused"
		}
		return "stopped"
	default:
		return protocol.ProxyRuntimeStateError
	}
}
